<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Preamble | RNA-Seq analysis with R and Bioconductor" />
<meta property="og:type" content="book" />




<meta name="author" content="Laurent Gatto and Axelle Loriot" />

<meta name="date" content="2023-10-04" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Preamble | RNA-Seq analysis with R and Bioconductor">

<title>Preamble | RNA-Seq analysis with R and Bioconductor</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>
<script src="libs/htmlwidgets-1.6.2/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.28/datatables.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="libs/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.13.4/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">RNA-Seq analysis with R and Bioconductor<p><p class="author">Laurent Gatto and Axelle Loriot</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a id="active-page" href="index.html" id="toc-preamble">Preamble</a><ul class="toc-sections">
<li class="toc"><a href="#installation-instructions">Installation instructions</a></li>
<li class="toc"><a href="#data-preparation">Data preparation</a></li>
<li class="toc"><a href="#references">References</a></li>
</ul>
<a href="finding-differentially-expressed-features.html" id="toc-finding-differentially-expressed-features"><span class="toc-section-number">1</span> Finding differentially expressed features</a>
<a href="sec-se.html" id="toc-sec-se"><span class="toc-section-number">2</span> SummarizedExperiments</a>
<a href="sec-dimred.html" id="toc-sec-dimred"><span class="toc-section-number">3</span> Dimensionality reduction</a>
<a href="sec-hts.html" id="toc-sec-hts"><span class="toc-section-number">4</span> High throughput sequencing</a>
<a href="sec-rnaseq.html" id="toc-sec-rnaseq"><span class="toc-section-number">5</span> Differential expression analysis</a>
<a href="sec-gsea.html" id="toc-sec-gsea"><span class="toc-section-number">6</span> Enrichment analyses</a>
<a href="conclusions.html" id="toc-conclusions"><span class="toc-section-number">7</span> Conclusions</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--bookdown:title:end--><!--bookdown:title:start--><html><body><div id="preamble" class="section level1 unnumbered">
<h1>Preamble</h1>
<p>The main objective of this course is to demonstrate how to proceed
with a complete analysis of RNA-seq data, including principal
component analysis, identification of differentially expressed genes
and gene set enrichment and over-representation analysis.</p>
<p>The background for these analyses inludes introductions to hypthesis
testing, linear modelling and adjustment for multiple comparisons,
principal component analysis and high throughtput sequencing.</p>
<p>This course is based on material from the
<a href="http://bit.ly/WSBIM1322">WSBIM1322</a> and
<a href="http://bit.ly/WSBIM2122">WSBIM2122</a> chapters taught at the <a href="https://uclouvain.be/en/faculties/fasb">Faculty
of Pharmacy and Biomedical
Sciences</a> at the UCLouvain.</p>
<div id="installation-instructions" class="section level2 unnumbered">
<h2>Installation instructions<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('installation-instructions')" onmouseout="reset_tooltip('installation-instructions-tooltip')"><span class="tooltiptext" id="installation-instructions-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="index.html#cb1-1" tabindex="-1"></a>pks <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"tidyverse"</span>, <span class="st">"patchwork"</span>, <span class="st">"remotes"</span>,</span>
<span id="cb1-2"><a href="index.html#cb1-2" tabindex="-1"></a>         <span class="st">"SummarizedExperiment"</span>, <span class="st">"DESeq2"</span>,</span>
<span id="cb1-3"><a href="index.html#cb1-3" tabindex="-1"></a>         <span class="st">"biomaRt"</span>, <span class="st">"factoextra"</span>, <span class="st">"pRolocdata"</span>,</span>
<span id="cb1-4"><a href="index.html#cb1-4" tabindex="-1"></a>         <span class="st">"org.Hs.eg.db"</span>, <span class="st">"GO.db"</span>, <span class="st">"msigdbr"</span>,</span>
<span id="cb1-5"><a href="index.html#cb1-5" tabindex="-1"></a>         <span class="st">"clusterProfiler"</span>, <span class="st">"airway"</span>,</span>
<span id="cb1-6"><a href="index.html#cb1-6" tabindex="-1"></a>         <span class="st">"UCLouvain-CBIO/rWSBIM2122"</span>,</span>
<span id="cb1-7"><a href="index.html#cb1-7" tabindex="-1"></a>         <span class="st">"UCLouvain-CBIO/rWSBIM1322"</span>)</span>
<span id="cb1-8"><a href="index.html#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="index.html#cb1-9" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>))</span>
<span id="cb1-10"><a href="index.html#cb1-10" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-11"><a href="index.html#cb1-11" tabindex="-1"></a></span>
<span id="cb1-12"><a href="index.html#cb1-12" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(pks)</span></code></pre></div>
</div>
<div id="data-preparation" class="section level2 unnumbered">
<h2>Data preparation<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('data-preparation')" onmouseout="reset_tooltip('data-preparation-tooltip')"><span class="tooltiptext" id="data-preparation-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>To prepare soem of the data needed in this course, run
<code>rWSBIM2122::prepare_shell()</code> as shown below:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="index.html#cb2-1" tabindex="-1"></a>rWSBIM2122<span class="sc">::</span><span class="fu">prepare_shell</span>()</span></code></pre></div>
<pre><code>## Preparing shell data in 'wsbim2122_data'.</code></pre>
<pre><code>## wsbim2122_data is ready.</code></pre>
<p>The function creates a new directory called <code>wsbim2122_data</code>
containing the following sub-directories:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="index.html#cb5-1" tabindex="-1"></a><span class="fu">dir</span>(<span class="st">"wsbim2122_data"</span>)</span></code></pre></div>
<pre><code>## [1] "count_data"      "data"            "deseq2"          "RNAseq_pipeline"</code></pre>
</div>
<div id="references" class="section level2 unnumbered">
<h2>References<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('references')" onmouseout="reset_tooltip('references-tooltip')"><span class="tooltiptext" id="references-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<ul>
<li><p>References are provided throughout the workshop material.</p></li>
<li><p><strong>Modern Statistics for Modern Biology</strong>, by Susan Holmes and
Wolfgang Huber <span class="citation">(<label for="tufte-mn-1" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-1" class="margin-toggle">Holmes and Huber 2019<span class="marginnote">Holmes, Susan, and Wolfgang Huber. 2019. <em>Modern Statistics for Modern Biology</em>. Cambridge Univeristy Press.</span>)</span>. A free online version of the book is
available <a href="https://www-huber.embl.de/msmb/">here</a>.</p></li>
</ul>
</div>
</div></body></html>

<p style="text-align: center;">
<a href="finding-differentially-expressed-features.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2023-10-04
 using 
R version 4.3.1 Patched (2023-07-10 r84676)
</p>
</div>
</div>



</body>
</html>
