<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 3 Dimensionality reduction | RNA-Seq analysis with R and Bioconductor" />
<meta property="og:type" content="book" />




<meta name="author" content="Laurent Gatto and Axelle Loriot" />

<meta name="date" content="2023-10-03" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Chapter 3 Dimensionality reduction | RNA-Seq analysis with R and Bioconductor">

<title>Chapter 3 Dimensionality reduction | RNA-Seq analysis with R and Bioconductor</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>
<script src="libs/htmlwidgets-1.6.2/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.28/datatables.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="libs/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.13.4/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">RNA-Seq analysis with R and Bioconductor<p><p class="author">Laurent Gatto and Axelle Loriot</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html" id="toc-preamble">Preamble</a>
<a href="finding-differentially-expressed-features.html" id="toc-finding-differentially-expressed-features"><span class="toc-section-number">1</span> Finding differentially expressed features</a>
<a href="sec-se.html" id="toc-sec-se"><span class="toc-section-number">2</span> SummarizedExperiments</a>
<a id="active-page" href="sec-dimred.html" id="toc-sec-dimred"><span class="toc-section-number">3</span> Dimensionality reduction</a><ul class="toc-sections">
<li class="toc"><a href="#introduction"> Introduction</a></li>
<li class="toc"><a href="#lower-dimensional-projections"> Lower-dimensional projections</a></li>
<li class="toc"><a href="#the-new-linear-combinations"> The new linear combinations</a></li>
<li class="toc"><a href="#summary-and-application"> Summary and application</a></li>
<li class="toc"><a href="#visualisation"> Visualisation</a></li>
<li class="toc"><a href="#pre-processing-and-missing-values-with-pca"> Pre-processing and missing values with PCA</a></li>
<li class="toc"><a href="#the-full-pca-workflow"> The full PCA workflow</a></li>
<li class="toc"><a href="#a-note-on-non-linear-dimensionality-reduction-techniques"> A note on non-linear dimensionality reduction techniques</a></li>
<li class="toc"><a href="#additional-exercises"> Additional exercises</a></li>
</ul>
<a href="sec-hts.html" id="toc-sec-hts"><span class="toc-section-number">4</span> High throughput sequencing</a>
<a href="sec-rnaseq.html" id="toc-sec-rnaseq"><span class="toc-section-number">5</span> Differential expression analysis</a>
<a href="sec-gsea.html" id="toc-sec-gsea"><span class="toc-section-number">6</span> Enrichment analyses</a>
<a href="conclusions.html" id="toc-conclusions"><span class="toc-section-number">7</span> Conclusions</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div id="sec-dimred" class="section level1" number="3">
<h1>
<span class="header-section-number">Chapter 3</span> Dimensionality reduction</h1>
<div id="introduction" class="section level2" number="3.1">
<h2>
<span class="header-section-number">3.1</span> Introduction<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('introduction')" onmouseout="reset_tooltip('introduction-tooltip')"><span class="tooltiptext" id="introduction-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>This chapter on PCA is partly based on the corresponding MSMB chapter
<span class="citation">(<label for="tufte-mn-2" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-2" class="margin-toggle">Holmes and Huber 2019<span class="marginnote">Holmes, Susan, and Wolfgang Huber. 2019. <em>Modern Statistics for Modern Biology</em>. Cambridge Univeristy Press.</span>)</span>. We are going to learn about dimensionality reduction, also
called ordination. The goal of dimensionality reduction is to
transform a high-dimensional data into data of lesser dimensions while
minimising the loss of information.</p>
<p>Dimensionality reduction is used as a data transformation technique
for input to other machine learning methods such as classifications,
or as a very efficient visualisation technique, which is the use case
we will focus on here. We are going to focus on a widerly used method
called Principal Component Analysis (PCA).</p>
<p>We are going to use the following dataset to illustrate some important
concepts that are scale and central to PCA. The small dataset show
below represents the measurement of genes <em>x</em> and <em>y</em> in 20
samples. We will be using the scaled and centred version of this data.</p>
<div class="figure">
<span style="display:block;" id="fig:xyplot"></span>
<p class="caption marginnote shownote">
Figure 3.1: Raw (left) and scale/centred (right) expression data for genes <em>x</em> and <em>y</em> in 20 samples
</p>
<img src="rnaseq_files/figure-html/xyplot-1.png" alt="Raw (left) and scale/centred (right) expression data for genes *x* and *y* in 20 samples" width="1056">
</div>
</div>
<div id="lower-dimensional-projections" class="section level2" number="3.2">
<h2>
<span class="header-section-number">3.2</span> Lower-dimensional projections<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('lower-dimensional-projections')" onmouseout="reset_tooltip('lower-dimensional-projections-tooltip')"><span class="tooltiptext" id="lower-dimensional-projections-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>The goal of dimensionality reduction is to reduce the number of
dimensions in a way that the new data remains useful. One way to
reduce a 2-dimensional data is by projecting the data onto a
lines. Below, we project our data on the x and y axes. These are
called <strong>linear projections</strong>.</p>
<div class="figure">
<span style="display:block;" id="fig:linproj"></span>
<p class="caption marginnote shownote">
Figure 3.2: Projection of the data on the x (left) and y (right) axes.
</p>
<img src="rnaseq_files/figure-html/linproj-1.png" alt="Projection of the data on the x (left) and y (right) axes." width="1056">
</div>
<p>In general, and in particular in the projections above, we lose
information when reducing the number of dimensions (above, from 2
(plane) to 1 (line)). In the first example above (left), we lose all
the measurements of gene <em>y</em>. In the second example (right), we lose
all the measurements of gene <em>x</em>.</p>
<p>The goal of dimensionality reduction is to limit this loss.</p>
<p>We know already about <strong>linear regression</strong>. Below, we use the <code>lm</code>
function to regress <em>y</em> onto <em>x</em> (left) and <em>x</em> onto <em>y</em>
(right). These regression lines give us an approximate linear
relationship between the expression of genes <em>x</em> and <em>y</em>. The
relationship differs depending on which gene we choose to be the
predictor and the response.</p>
<div class="figure">
<span style="display:block;" id="fig:linreg"></span>
<p class="caption marginnote shownote">
Figure 3.3: Regression of y onto x (left) minimisises the sums of squares of vertical residuals (red). Regression of x onto y (right) minimisises the sums of squares of horizontal residuals (orange).
</p>
<img src="rnaseq_files/figure-html/linreg-1.png" alt="Regression of y onto x (left) minimisises the sums of squares of vertical residuals (red). Regression of x onto y (right) minimisises the sums of squares of horizontal residuals (orange)." width="1056">
</div>
<p>We now want a line that minimises distances in both directions, as
shown below. This line, called <strong>principal component</strong>, is also the
ones that maximises the variance of the projections along itself.</p>
<div class="figure">
<span style="display:block;" id="fig:pxaex"></span>
<p class="caption marginnote shownote">
Figure 3.4: The first prinicpal component minimises the sum of squares of the orthogonal projections.
</p>
<img src="rnaseq_files/figure-html/pxaex-1.png" alt="The first prinicpal component minimises the sum of squares of the orthogonal projections." width="1104">
</div>
<p>The second principal component is then chosen to be orthogonal to the
first one. In our case above, there is only one possibility.</p>
<div class="figure">
<span style="display:block;" id="fig:pxaex2"></span>
<p class="caption marginnote shownote">
Figure 3.5: The second prinicpal component is orthogonal to the second one.
</p>
<img src="rnaseq_files/figure-html/pxaex2-1.png" alt="The second prinicpal component is orthogonal to the second one." width="1056">
</div>
<p>In the example above the variance, the variance along the PCs are
1.77 and 0.23. The first one explains
88.6% or that variance, and the second one merely
11.4%. This is also reflected in the different
scales along the x and y axis.</p>
<p>To account for these differences in variation along the different PCs,
it is better to represent a PCA plot as a rectangle, using an aspect
ratio that is illustrative of the respective variances.</p>
<div class="figure">
<span style="display:block;" id="fig:pxaex3"></span>
<p class="caption marginnote shownote">
Figure 3.6: Final principal component analysis of the data.
</p>
<img src="rnaseq_files/figure-html/pxaex3-1.png" alt="Final principal component analysis of the data." width="960">
</div>
</div>
<div id="the-new-linear-combinations" class="section level2" number="3.3">
<h2>
<span class="header-section-number">3.3</span> The new linear combinations<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('the-new-linear-combinations')" onmouseout="reset_tooltip('the-new-linear-combinations-tooltip')"><span class="tooltiptext" id="the-new-linear-combinations-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Principal components are linear combinations of the variables that
were originally measured, they provide a new coordinate system. The
PC in the previous example is a linear combination of <em>gene_x</em> and
<em>gene_y</em>, more specifically</p>
<p><span class="math display">\[ PC = c_{1} ~ gene_{x} + c_{2} ~ gene_{y} \]</span></p>
<p>It has coefficients <span class="math inline">\((c_1, c_2)\)</span>, also called loading.</p>
<p>PCA in general will find linear combinations of the original
variables. These new linear combinations will maximise the variance
of the data.</p>
</div>
<div id="summary-and-application" class="section level2" number="3.4">
<h2>
<span class="header-section-number">3.4</span> Summary and application<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('summary-and-application')" onmouseout="reset_tooltip('summary-and-application-tooltip')"><span class="tooltiptext" id="summary-and-application-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Principal Component Analysis (PCA) is a technique that transforms the
original n-dimensional data into a new space.</p>
<ul>
<li><p>These new dimensions are linear combinations of the original data,
i.e. they are composed of proportions of the original variables.</p></li>
<li><p>Along these new dimensions, called principal components, the data
expresses most of its variability along the first PC, then second,
…</p></li>
<li><p>Principal components are orthogonal to each other,
i.e. non-correlated.</p></li>
</ul>
<div class="figure fullwidth">
<span style="display:block;" id="fig:pcaex"></span>
<img src="rnaseq_files/figure-html/pcaex-1.png" alt="Original data (left). PC1 will maximise the variability while minimising the residuals (centre). PC2 is orthogonal to PC1 (right)." width="1152"><p class="caption marginnote shownote">
Figure 3.7: Original data (left). PC1 will maximise the variability while minimising the residuals (centre). PC2 is orthogonal to PC1 (right).
</p>
</div>
<p>In R, we can use the <code>prcomp</code> function. A summary of the <code>prcomp</code>
output shows that along PC1, we are able to retain close to 92%
of the total variability in the data.</p>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="sec-dimred.html#cb156-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rWSBIM1322"</span>)</span>
<span id="cb156-2"><a href="sec-dimred.html#cb156-2" tabindex="-1"></a><span class="fu">data</span>(xy)</span>
<span id="cb156-3"><a href="sec-dimred.html#cb156-3" tabindex="-1"></a>pca_xy <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(xy)</span>
<span id="cb156-4"><a href="sec-dimred.html#cb156-4" tabindex="-1"></a><span class="fu">summary</span>(pca_xy)</span></code></pre></div>
<pre><code>## Importance of components:
##                           PC1    PC2
## Standard deviation     1.3308 0.4784
## Proportion of Variance 0.8856 0.1144
## Cumulative Proportion  0.8856 1.0000</code></pre>
<p>This <code>pca_xy</code> variable is an object of class <code>prcomp</code>. To learn what
it contains, we can look at its structure with <code>str</code> and read the
<code>?prcomp</code> manual page.</p>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="sec-dimred.html#cb158-1" tabindex="-1"></a><span class="fu">str</span>(pca_xy)</span></code></pre></div>
<pre><code>## List of 5
##  $ sdev    : num [1:2] 1.331 0.478
##  $ rotation: num [1:2, 1:2] 0.707 0.707 -0.707 0.707
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : chr [1:2] "gene_x" "gene_y"
##   .. ..$ : chr [1:2] "PC1" "PC2"
##  $ center  : Named num [1:2] 4.05e-16 1.53e-17
##   ..- attr(*, "names")= chr [1:2] "gene_x" "gene_y"
##  $ scale   : logi FALSE
##  $ x       : num [1:20, 1:2] -2.728 -2.4 -2.257 -0.945 -0.14 ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : NULL
##   .. ..$ : chr [1:2] "PC1" "PC2"
##  - attr(*, "class")= chr "prcomp"</code></pre>
<p>We are going to focus on two elements:</p>
<ul>
<li>
<strong>sdev</strong> contains the standard deviations along the respective PCs
(as also displayed in the summary). From these, we can compute the
variances, the percentages of variance explained by the individual
PCs, and the cumulative variances.</li>
</ul>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="sec-dimred.html#cb160-1" tabindex="-1"></a>(var <span class="ot">&lt;-</span> pca_xy<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>)</span></code></pre></div>
<pre><code>## [1] 1.7711407 0.2288593</code></pre>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="sec-dimred.html#cb162-1" tabindex="-1"></a>(pve <span class="ot">&lt;-</span> var<span class="sc">/</span><span class="fu">sum</span>(var))</span></code></pre></div>
<pre><code>## [1] 0.8855704 0.1144296</code></pre>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="sec-dimred.html#cb164-1" tabindex="-1"></a><span class="fu">cumsum</span>(pve)</span></code></pre></div>
<pre><code>## [1] 0.8855704 1.0000000</code></pre>
<ul>
<li>
<strong>x</strong> contains the coordinates of the data along the PCs. These are
the values we could use to produce the PCA plot as above by hand.</li>
</ul>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="sec-dimred.html#cb166-1" tabindex="-1"></a>pca_xy<span class="sc">$</span>x</span></code></pre></div>
<pre><code>##               PC1         PC2
##  [1,] -2.72848772  0.17019003
##  [2,] -2.40018291  0.11371158
##  [3,] -2.25654153 -0.17987834
##  [4,] -0.94491834  0.14224604
##  [5,] -0.14044618  0.86760695
##  [6,] -0.85917747  0.06485259
##  [7,] -1.26714805 -0.38570171
##  [8,] -0.13426336  0.01962124
##  [9,]  0.14678704  0.14527416
## [10,]  1.08677349  0.67252884
## [11,]  0.93793985  0.43069778
## [12,]  0.62416628  0.08630825
## [13,]  0.05477564 -0.56143151
## [14,]  1.20838962  0.46257985
## [15,]  0.58230063 -0.44200381
## [16,]  1.48091889  0.42011071
## [17,]  1.13374563 -0.12764440
## [18,]  1.28218805 -0.32631519
## [19,]  1.46938699 -0.32768480
## [20,]  0.72379344 -1.24506826</code></pre>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="sec-dimred.html#cb168-1" tabindex="-1"></a><span class="fu">plot</span>(pca_xy<span class="sc">$</span>x)</span></code></pre></div>
<p><img src="rnaseq_files/figure-html/pcaplot-1.png" width="672"></p>
</div>
<div id="visualisation" class="section level2" number="3.5">
<h2>
<span class="header-section-number">3.5</span> Visualisation<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('visualisation')" onmouseout="reset_tooltip('visualisation-tooltip')"><span class="tooltiptext" id="visualisation-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>A <strong>biplot</strong> features all original points re-mapped (rotated) along the
first two PCs as well as the original features as vectors along the
same PCs. Feature vectors that are in the same direction in PC space
are also correlated in the original data space.</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="sec-dimred.html#cb169-1" tabindex="-1"></a><span class="fu">biplot</span>(pca_xy)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:biplot"></span>
<p class="caption marginnote shownote">
Figure 3.8: A biplot shows both the variables (arrows) and observations of the PCA analysis.
</p>
<img src="rnaseq_files/figure-html/biplot-1.png" alt="A biplot shows both the variables (arrows) and observations of the PCA analysis." width="672">
</div>
<p>One important piece of information when using PCA is the proportion of
variance explained along the PCs (see above), in particular when
dealing with high dimensional data, as PC1 and PC2 (that are generally
used for visualisation), might only account for an insufficient
proportion of variance to be relevant on their own. This can be
visualised on a <strong>screeplot</strong>, that can be produced with</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="sec-dimred.html#cb170-1" tabindex="-1"></a><span class="fu">plot</span>(pca_xy)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:screeplot"></span>
<p class="caption marginnote shownote">
Figure 3.9: Screeplot showing the variances for the PCs.
</p>
<img src="rnaseq_files/figure-html/screeplot-1.png" alt="Screeplot showing the variances for the PCs." width="672">
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Load the <code>cptac_se_prot</code> data available in the <code>rWSBIM1322</code> package,
scale the data (see below for why scaling is important), then perform
the PCA analysis and interpret it. Also produce a PCA for PCs 2 and 3.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-10" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-10', 'sol-start-10')"></span>
</p>
<div id="sol-body-10" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="sec-dimred.html#cb171-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rWSBIM1322"</span>)</span>
<span id="cb171-2"><a href="sec-dimred.html#cb171-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"SummarizedExperiment"</span>)</span>
<span id="cb171-3"><a href="sec-dimred.html#cb171-3" tabindex="-1"></a><span class="fu">data</span>(cptac_se_prot)</span>
<span id="cb171-4"><a href="sec-dimred.html#cb171-4" tabindex="-1"></a>cptac_se_prot</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 114 6 
## metadata(3): MSnbaseFiles MSnbaseProcessing MSnbaseVersion
## assays(1): ''
## rownames(114): P00918ups|CAH2_HUMAN_UPS
##   P01008ups|ANT3_HUMAN_UPS;CON__P41361 ... sp|P48837|NUP57_YEAST
##   sp|P47912|LCF4_YEAST;sp|P39002|LCF3_YEAST
## rowData names(3): Proteins Sequence nNA
## colnames(6): 6A_7 6A_8 ... 6B_8 6B_9
## colData names(3): group sample nNA</code></pre>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="sec-dimred.html#cb173-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">scale</span>(<span class="fu">assay</span>(cptac_se_prot))</span>
<span id="cb173-2"><a href="sec-dimred.html#cb173-2" tabindex="-1"></a>pca2 <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(x)</span>
<span id="cb173-3"><a href="sec-dimred.html#cb173-3" tabindex="-1"></a><span class="fu">summary</span>(pca2)</span></code></pre></div>
<pre><code>## Importance of components:
##                           PC1    PC2     PC3     PC4     PC5     PC6
## Standard deviation     2.3718 0.4106 0.30777 0.22335 0.20650 0.13692
## Proportion of Variance 0.9376 0.0281 0.01579 0.00831 0.00711 0.00312
## Cumulative Proportion  0.9376 0.9657 0.98145 0.98977 0.99688 1.00000</code></pre>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="sec-dimred.html#cb175-1" tabindex="-1"></a><span class="fu">screeplot</span>(pca2)</span></code></pre></div>
<p><img src="rnaseq_files/figure-html/pcaex2-1.png" width="672"></p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="sec-dimred.html#cb176-1" tabindex="-1"></a><span class="fu">biplot</span>(pca2)</span></code></pre></div>
<p><img src="rnaseq_files/figure-html/pcaex2-2.png" width="672"></p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="sec-dimred.html#cb177-1" tabindex="-1"></a><span class="fu">plot</span>(pca2<span class="sc">$</span>x[, <span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>])</span></code></pre></div>
<p><img src="rnaseq_files/figure-html/pcaex2-3.png" width="672"></p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="sec-dimred.html#cb178-1" tabindex="-1"></a><span class="fu">biplot</span>(pca2, <span class="at">choices =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="rnaseq_files/figure-html/pcaex2-4.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>In the exercise above, the PCA was performed on the features
(proteins). Transpose the data and produce a PCA of the
samples. Interprete the figures.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-11" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-11', 'sol-start-11')"></span>
</p>
<div id="sol-body-11" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="sec-dimred.html#cb179-1" tabindex="-1"></a>x_t <span class="ot">&lt;-</span> <span class="fu">t</span>(x)</span>
<span id="cb179-2"><a href="sec-dimred.html#cb179-2" tabindex="-1"></a>pca3 <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(x_t)</span>
<span id="cb179-3"><a href="sec-dimred.html#cb179-3" tabindex="-1"></a><span class="fu">summary</span>(pca3)</span></code></pre></div>
<pre><code>## Importance of components:
##                           PC1    PC2    PC3    PC4     PC5       PC6
## Standard deviation     1.9522 1.4632 1.0618 0.9817 0.65186 8.723e-16
## Proportion of Variance 0.4501 0.2528 0.1331 0.1138 0.05018 0.000e+00
## Cumulative Proportion  0.4501 0.7029 0.8360 0.9498 1.00000 1.000e+00</code></pre>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="sec-dimred.html#cb181-1" tabindex="-1"></a><span class="fu">screeplot</span>(pca3)</span></code></pre></div>
<p><img src="rnaseq_files/figure-html/unnamed-chunk-45-1.png" width="672"></p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="sec-dimred.html#cb182-1" tabindex="-1"></a><span class="fu">biplot</span>(pca3)</span></code></pre></div>
<p><img src="rnaseq_files/figure-html/unnamed-chunk-45-2.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="pre-processing-and-missing-values-with-pca" class="section level2" number="3.6">
<h2>
<span class="header-section-number">3.6</span> Pre-processing and missing values with PCA<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('pre-processing-and-missing-values-with-pca')" onmouseout="reset_tooltip('pre-processing-and-missing-values-with-pca-tooltip')"><span class="tooltiptext" id="pre-processing-and-missing-values-with-pca-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>We haven’t looked at other <code>prcomp</code> parameters, other that the first
one, <code>x</code>. There are two other ones that are or importance, in
particular in the light of the section on pre-processing above, which
are <code>center</code> and <code>scale.</code>. The former is set to <code>TRUE</code> by default,
while the second one is set the <code>FALSE</code>.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Perform a PCA analysis on the <code>mtcars</code> dataset with and without
scaling. Compare and interpret the results.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-12" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-12', 'sol-start-12')"></span>
</p>
<div id="sol-body-12" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="sec-dimred.html#cb183-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb183-2"><a href="sec-dimred.html#cb183-2" tabindex="-1"></a><span class="fu">biplot</span>(<span class="fu">prcomp</span>(mtcars, <span class="at">scale =</span> <span class="cn">FALSE</span>), <span class="at">main =</span> <span class="st">"No scaling"</span>)  <span class="do">## 1</span></span>
<span id="cb183-3"><a href="sec-dimred.html#cb183-3" tabindex="-1"></a><span class="fu">biplot</span>(<span class="fu">prcomp</span>(mtcars, <span class="at">scale =</span> <span class="cn">TRUE</span>), <span class="at">main =</span> <span class="st">"With scaling"</span>) <span class="do">## 2</span></span></code></pre></div>
<p><img src="rnaseq_files/figure-html/scalepcasol-1.png" width="672"></p>
<p>Without scaling, <code>disp</code> and <code>hp</code> are the features with the highest
loadings along PC1 and 2 (all others are negligible), which are also
those with the highest units of measurement. Scaling removes this
effect.</p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Real datasets often come with <em>missing values</em>. In R, these should be
encoded using <code>NA</code>. Unfortunately, PCA cannot deal with missing
values, and observations containing <code>NA</code> values will be dropped
automatically. This is a viable solution only when the proportion of
missing values is low. Otherwise, it is possible to impute missing
values (which often requires great care) or use an implementation of
PCA such as <em>non-linear iterative partial least squares</em> (NIPALS),
that support missing values.</p>
</div>
<div id="the-full-pca-workflow" class="section level2" number="3.7">
<h2>
<span class="header-section-number">3.7</span> The full PCA workflow<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('the-full-pca-workflow')" onmouseout="reset_tooltip('the-full-pca-workflow-tooltip')"><span class="tooltiptext" id="the-full-pca-workflow-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>In this section, we will describe the detailed PCA analysis and
interpretation of two real-life proteomics datasets. Both of these
datasets are available in the <code>pRolocdata</code> package.</p>
<p>We are first going to focus on the data from Mulvey <em>et al.</em> (2015),
where they present the <em>Dynamic proteomic profiling of extra-embryonic
endoderm differentiation in mouse embryonic stem cells</em> <span class="citation">(<label for="tufte-mn-3" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-3" class="margin-toggle">Mulvey et al. 2015<span class="marginnote">Mulvey, C M, C Schröter, L Gatto, D Dikicioglu, I Baris Fidaner, A Christoforou, M J Deery, et al. 2015. <span>“Dynamic Proteomic Profiling of Extra-Embryonic Endoderm Differentiation in Mouse Embryonic Stem Cells.”</span> <em>Stem Cells</em>, June. <a href="https://doi.org/10.1002/stem.2067">https://doi.org/10.1002/stem.2067</a>.</span>)</span>.</p>
<blockquote>
<p>During mammalian preimplantation development, the cells of the
blastocyst’s inner cell mass differentiate into the epiblast and
primitive endoderm lineages, which give rise to the fetus and
extra-embryonic tissues, respectively. Extra-embryonic endoderm
(XEN) differentiation can be modeled in vitro by induced
expression of GATA transcription factors in mouse embryonic stem
cells. Here, we use this GATA-inducible system to quantitatively
monitor the dynamics of global proteomic changes during the early
stages of this differentiation event and also investigate the
fully differentiated phenotype, as represented by embryo-derived
XEN cells. Using mass spectrometry-based quantitative proteomic
profiling with multivariate data analysis tools, we reproducibly
quantified 2,336 proteins across three biological replicates and
have identified clusters of proteins characterized by distinct,
dynamic temporal abundance profiles. We first used this approach
to highlight novel marker candidates of the pluripotent state and
XEN differentiation. Through functional annotation enrichment
analysis, we have shown that the downregulation of
chromatin-modifying enzymes, the reorganization of membrane
trafficking machinery, and the breakdown of cell-cell adhesion are
successive steps of the extra-embryonic differentiation process.
Thus, applying a range of sophisticated clustering approaches to a
time-resolved proteomic dataset has allowed the elucidation of
complex biological processes which characterize stem cell
differentiation and could establish a general paradigm for the
investigation of these processes.</p>
</blockquote>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Load the <code>mulvey2015_se</code> data from the <code>pRolocdata</code> package. The data
comes as an object of class <code>SummarizedExperiment</code>. Familiarise
yourself with its experimental design stored in the <em>colData</em> slot.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-13" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-13', 'sol-start-13')"></span>
</p>
<div id="sol-body-13" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="sec-dimred.html#cb184-1" tabindex="-1"></a><span class="fu">library</span>(pRolocdata)</span>
<span id="cb184-2"><a href="sec-dimred.html#cb184-2" tabindex="-1"></a><span class="fu">data</span>(mulvey2015_se)</span>
<span id="cb184-3"><a href="sec-dimred.html#cb184-3" tabindex="-1"></a><span class="fu">colData</span>(mulvey2015_se)</span></code></pre></div>
<pre><code>## DataFrame with 18 rows and 3 columns
##                 rep     times      cond
##           &lt;numeric&gt; &lt;integer&gt; &lt;numeric&gt;
## rep1_0hr          1         1         1
## rep1_16hr         1         2         1
## rep1_24hr         1         3         1
## rep1_48hr         1         4         1
## rep1_72hr         1         5         1
## ...             ...       ...       ...
## rep3_16hr         3         2         1
## rep3_24hr         3         3         1
## rep3_48hr         3         4         1
## rep3_72hr         3         5         1
## rep3_XEN          3         6         1</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Here, we will want to do a PCA analysis on the samples. We want to
remap the 18 samples from a 2337-dimensional
space into 2, or possibly 3 dimensions. This will require to transpose
the data matrix before passing it to the <code>prcomp</code> function.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li>Run a PCA analysis on the samples of the <code>mulvey2015_se</code> data and
display a summary of the results.</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-14" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-14', 'sol-start-14')"></span>
</p>
<div id="sol-body-14" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="sec-dimred.html#cb186-1" tabindex="-1"></a>pca_mulvey2015 <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(<span class="fu">assay</span>(mulvey2015_se)), <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span>
<span id="cb186-2"><a href="sec-dimred.html#cb186-2" tabindex="-1"></a><span class="fu">summary</span>(pca_mulvey2015)</span></code></pre></div>
<pre><code>## Importance of components:
##                            PC1     PC2     PC3     PC4     PC5     PC6     PC7
## Standard deviation     34.2468 24.0808 16.8485 7.37212 6.79981 5.84729 5.50515
## Proportion of Variance  0.5019  0.2481  0.1215 0.02326 0.01978 0.01463 0.01297
## Cumulative Proportion   0.5019  0.7500  0.8715 0.89472 0.91450 0.92913 0.94210
##                            PC8    PC9    PC10    PC11    PC12    PC13    PC14
## Standard deviation     5.18188 4.7114 4.47518 3.99894 3.88066 3.52931 3.24782
## Proportion of Variance 0.01149 0.0095 0.00857 0.00684 0.00644 0.00533 0.00451
## Cumulative Proportion  0.95359 0.9631 0.97166 0.97850 0.98494 0.99027 0.99479
##                           PC15    PC16   PC17      PC18
## Standard deviation     3.04106 1.24069 1.1806 1.709e-14
## Proportion of Variance 0.00396 0.00066 0.0006 0.000e+00
## Cumulative Proportion  0.99874 0.99940 1.0000 1.000e+00</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li><p>Assuming we are happy with a reduced space explaining 90% of the
variance of the data, how many PCs do we need?</p></li>
<li><p>Visualise the variance captured by the 18 PCs. To do so, you can use
the <code>fviz_screeplot</code> function from the <code>factoextra</code> package.</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-15" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-15', 'sol-start-15')"></span>
</p>
<div id="sol-body-15" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="sec-dimred.html#cb188-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"factoextra"</span>)</span>
<span id="cb188-2"><a href="sec-dimred.html#cb188-2" tabindex="-1"></a><span class="fu">fviz_screeplot</span>(pca_mulvey2015)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-48"></span>
<p class="caption marginnote shownote">
Figure 3.10: Scree plot for the <code>mulvey2015_se</code> data showing the percentage of explained variance onlong the principal components.
</p>
<img src="rnaseq_files/figure-html/unnamed-chunk-48-1.png" alt="Scree plot for the `mulvey2015_se` data showing the percentage of explained variance onlong the principal components." width="672">
</div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li><p>Visualise your PCA analysis on a biplot. You can use the <code>fviz_pca</code>
function from the <code>factoextra</code> package, or the <code>fviz_pca_ind</code>
function to focus on the individuals (the rows of the original
input).</p></li>
<li><p>Use the <code>habillage</code> argument of the <code>fviz_pca_ind</code> function to
highlight the time experimental variable. Interpret the figures in
the light of the experimental design.</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-16" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-16', 'sol-start-16')"></span>
</p>
<div id="sol-body-16" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="sec-dimred.html#cb189-1" tabindex="-1"></a><span class="fu">fviz_pca</span>(pca_mulvey2015)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-49-1"></span>
<p class="caption marginnote shownote">
Figure 3.11: Biplot and variables PCA plot of the <code>mulvey2015_se</code> data.
</p>
<img src="rnaseq_files/figure-html/unnamed-chunk-49-1.png" alt="Biplot and variables PCA plot of the `mulvey2015_se` data." width="768">
</div>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="sec-dimred.html#cb190-1" tabindex="-1"></a><span class="fu">fviz_pca_ind</span>(pca_mulvey2015)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-49-2"></span>
<p class="caption marginnote shownote">
Figure 3.12: Biplot and variables PCA plot of the <code>mulvey2015_se</code> data.
</p>
<img src="rnaseq_files/figure-html/unnamed-chunk-49-2.png" alt="Biplot and variables PCA plot of the `mulvey2015_se` data." width="768">
</div>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="sec-dimred.html#cb191-1" tabindex="-1"></a><span class="fu">fviz_pca_ind</span>(pca_mulvey2015, <span class="at">habillage =</span> mulvey2015_se<span class="sc">$</span>times)</span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-49-3"></span>
<p class="caption marginnote shownote">
Figure 3.13: Biplot and variables PCA plot of the <code>mulvey2015_se</code> data.
</p>
<img src="rnaseq_files/figure-html/unnamed-chunk-49-3.png" alt="Biplot and variables PCA plot of the `mulvey2015_se` data." width="768">
</div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>To visualise the data in 3 dimension, you can use the <code>plot3d</code>
function from the <code>rgl</code> package, that provides means to rotate the
cube along the three first PCs.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-17" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-17', 'sol-start-17')"></span>
</p>
<div id="sol-body-17" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="sec-dimred.html#cb192-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rgl"</span>)</span>
<span id="cb192-2"><a href="sec-dimred.html#cb192-2" tabindex="-1"></a><span class="fu">plot3d</span>(pca_mulvey2015<span class="sc">$</span>x[, <span class="dv">1</span>],</span>
<span id="cb192-3"><a href="sec-dimred.html#cb192-3" tabindex="-1"></a>       pca_mulvey2015<span class="sc">$</span>x[, <span class="dv">2</span>],</span>
<span id="cb192-4"><a href="sec-dimred.html#cb192-4" tabindex="-1"></a>       pca_mulvey2015<span class="sc">$</span>x[, <span class="dv">3</span>],</span>
<span id="cb192-5"><a href="sec-dimred.html#cb192-5" tabindex="-1"></a>       <span class="at">col =</span> mulvey2015_se<span class="sc">$</span>times,</span>
<span id="cb192-6"><a href="sec-dimred.html#cb192-6" tabindex="-1"></a>       <span class="at">size =</span> <span class="dv">10</span>)</span></code></pre></div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>See <a href="https://uclouvain-cbio.github.io/WSBIM1322/sec-dimred.html#the-full-pca-workflow">this
section</a>
for an another application on features this time, taken from the paper
‘<a href="https://dx.doi.org/10.1038/ncomms9992">A map of the mouse pluripotent stem cell spatial
proteome</a>’
<span class="citation">(<label for="tufte-mn-4" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-4" class="margin-toggle">Christoforou et al. 2016<span class="marginnote">Christoforou, A, C M Mulvey, L M Breckels, A Geladaki, T Hurrell, P C Hayward, T Naake, et al. 2016. <span>“A Draft Map of the Mouse Pluripotent Stem Cell Spatial Proteome.”</span> <em>Nat Commun</em> 7: 8992. <a href="https://doi.org/10.1038/ncomms9992">https://doi.org/10.1038/ncomms9992</a>.</span>)</span>.</p>
</div>
<div id="a-note-on-non-linear-dimensionality-reduction-techniques" class="section level2" number="3.8">
<h2>
<span class="header-section-number">3.8</span> A note on non-linear dimensionality reduction techniques<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('a-note-on-non-linear-dimensionality-reduction-techniques')" onmouseout="reset_tooltip('a-note-on-non-linear-dimensionality-reduction-techniques-tooltip')"><span class="tooltiptext" id="a-note-on-non-linear-dimensionality-reduction-techniques-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Non-linear dimensionality reduction techniques such as t-SNE or UMAP
are widely used, especially since the advent of single-cell RNA
sequencing. They do not aim at reducing the dimensions according to
variance, as PCA, but focus on preserve neighbourhood between samples
in a lower-dimensions. However, this neighbourhood is only maintained
for close samples: close points in the lower dimension are also close
in high dimensions, while it isn’t possible to make any conclusion for
distant points in low dimensions.</p>
</div>
<div id="additional-exercises" class="section level2" number="3.9">
<h2>
<span class="header-section-number">3.9</span> Additional exercises<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('additional-exercises')" onmouseout="reset_tooltip('additional-exercises-tooltip')"><span class="tooltiptext" id="additional-exercises-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Professor Simpson and her team wanted to test the effect of drug A and
drug B on fibroblast. They measured gene expression using RNA-Seq in
three conditions: a control without any drug, cells in presence of
drug A, and cells in presence of drug B. Each of these were measured
in triplicates in a single batch by the same operator.</p>
<p>The figure below shows the PCA plot of the 9 samples, coloured by
condition. Interpreted the figure in the light of the experimentally
design and suggest how to best analyse these data.</p>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-51"></span>
<p class="caption marginnote shownote">
Figure 3.14: PCA analysis illustrating the global effect of drug A and B.
</p>
<img src="rnaseq_files/figure-html/unnamed-chunk-51-1.png" alt="PCA analysis illustrating the global effect of drug A and B." width="672">
</div>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Consider the same experimental design as above, but with the PCA plot
below. Interpret the figure in the light of the experimental design
and suggest how to best analyse these data.</p>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-52"></span>
<p class="caption marginnote shownote">
Figure 3.15: PCA analysis illustrating the global effect of drug A and B.
</p>
<img src="rnaseq_files/figure-html/unnamed-chunk-52-1.png" alt="PCA analysis illustrating the global effect of drug A and B." width="672">
</div>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Researchers of the famous FLVIT (Fosse-La-Ville Institute of
Technology) wanted to test the effect of a newly identified
anti-cancer drug C121 on KEK293 cells. They measured proteins
expression using quantitative mass spectrometry in these two
conditions and 4 replicates. Each of these were measured in
triplicates in a single batch.</p>
<p>Interpret the PCA plot in the light of the experimental design. Do you
think the analysis is likely to provide significant results?</p>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-53"></span>
<p class="caption marginnote shownote">
Figure 3.16: Quantitative proteomics experiment measuring the effect of drug C121.
</p>
<img src="rnaseq_files/figure-html/unnamed-chunk-53-1.png" alt="Quantitative proteomics experiment measuring the effect of drug C121." width="672">
</div>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>In the following experimental design, clinicians have used
quantitative PCR to measure the expression of 96 micro RNAs in the
blood of three patients and three healthy donor, with the goal of
identifying novel biomarkers.</p>
<p>In the light of the experimental design, interpret the PCA plot. What
are the chances to find good biomarkers?</p>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-54"></span>
<p class="caption marginnote shownote">
Figure 3.17: Quantitative proteomics experiment measuring the effect of drug C121.
</p>
<img src="rnaseq_files/figure-html/unnamed-chunk-54-1.png" alt="Quantitative proteomics experiment measuring the effect of drug C121." width="672">
</div>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>The team of Professor Von der Brille have quantified the gene
expression of Jurkat cells using two RNA-Seq protocols. 25 samples
were tested with kit 1 by operator A and B (12 and 13 samples
respectively), and 27 samples were processed with kit 2 by operators A
and C (20 and 7 samples respectively).</p>
<p>In the light of the experimental design, interpret the PCA plot. What
would you suggest to do before testing whether using kit 1 and 2 have
an impact on the data?</p>
<div class="figure">
<span style="display:block;" id="fig:unnamed-chunk-55"></span>
<p class="caption marginnote shownote">
Figure 3.18: Testing the effect of RNA sequencing kits 1 and 2.
</p>
<img src="rnaseq_files/figure-html/unnamed-chunk-55-1.png" alt="Testing the effect of RNA sequencing kits 1 and 2." width="672">
</div>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>

</div>
</div></body></html>

<p style="text-align: center;">
<a href="sec-se.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-hts.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2023-10-03
 using 
R version 4.3.1 Patched (2023-07-10 r84676)
</p>
</div>
</div>



</body>
</html>
