<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 2 SummarizedExperiments | RNA-Seq analysis with R and Bioconductor" />
<meta property="og:type" content="book" />




<meta name="author" content="Laurent Gatto and Axelle Loriot" />

<meta name="date" content="2023-10-03" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Chapter 2 SummarizedExperiments | RNA-Seq analysis with R and Bioconductor">

<title>Chapter 2 SummarizedExperiments | RNA-Seq analysis with R and Bioconductor</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script>
function copy_link(id) {
  var dummy = document.createElement('input'),
  text = window.location.href.split(/[?#]/)[0] + '#' + id;
  document.body.appendChild(dummy);
  dummy.value = text;
  dummy.select();
  document.execCommand('copy');
  document.body.removeChild(dummy);
  
  var tooltip = document.getElementById(id + '-tooltip');
  tooltip.innerHTML = 'Copied!';
}

function reset_tooltip(id) {
  var tooltip = document.getElementById(id);
  tooltip.innerHTML = 'Copy link';
}
</script>
<script src="libs/htmlwidgets-1.6.2/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.28/datatables.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="libs/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.13.4/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }

code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">RNA-Seq analysis with R and Bioconductor<p><p class="author">Laurent Gatto and Axelle Loriot</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html" id="toc-preamble">Preamble</a>
<a href="finding-differentially-expressed-features.html" id="toc-finding-differentially-expressed-features"><span class="toc-section-number">1</span> Finding differentially expressed features</a>
<a id="active-page" href="sec-se.html" id="toc-sec-se"><span class="toc-section-number">2</span> SummarizedExperiments</a><ul class="toc-sections">
<li class="toc"><a href="#summarizedexperiment"> SummarizedExperiment</a></li>
<li class="toc"><a href="#rangedsummarizedexperiment"> RangedSummarizedExperiment</a></li>
<li class="toc"><a href="#creating-a-count-matrix-from-multiple-files"> Creating a count matrix from multiple files</a></li>
</ul>
<a href="sec-dimred.html" id="toc-sec-dimred"><span class="toc-section-number">3</span> Dimensionality reduction</a>
<a href="sec-hts.html" id="toc-sec-hts"><span class="toc-section-number">4</span> High throughput sequencing</a>
<a href="sec-rnaseq.html" id="toc-sec-rnaseq"><span class="toc-section-number">5</span> Differential expression analysis</a>
<a href="sec-gsea.html" id="toc-sec-gsea"><span class="toc-section-number">6</span> Enrichment analyses</a>
<a href="conclusions.html" id="toc-conclusions"><span class="toc-section-number">7</span> Conclusions</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body>
<div id="sec-se" class="section level1" number="2">
<h1>
<span class="header-section-number">Chapter 2</span> SummarizedExperiments</h1>
<p><strong>Objectives</strong></p>
<ul>
<li>Present the <code>SummarizedExperiment</code> class, extensively used in omics
analyses.</li>
<li>Show how to build <code>SummarizedExperiment</code> objects.</li>
</ul>
<p>Data in bioinformatics is often more complex than the basic data types
we have seen so far. To deal with this, developers define specialised
data containers (termed classes) that match the properties of the data
they need to handle.</p>
<p>This aspect is central to the <strong>Bioconductor</strong><a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a> project
which uses the same <strong>core data infrastructure</strong> across packages. This
certainly contributed to Bioconductor’s success. Bioconductor package
developers are advised to make use of <a href="https://uclouvain-cbio.github.io/WSBIM1322/sec-obj.html#examples-from-the-bioconductor-project">existing
infrastructure</a>
to provide coherence, interoperability and stability to the project as
a whole.</p>
<p>To illustrate such an omics data container, we’ll present the
<code>SummarizedExperiment</code> class.</p>
<div id="summarizedexperiment" class="section level2" number="2.1">
<h2>
<span class="header-section-number">2.1</span> SummarizedExperiment<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('summarizedexperiment')" onmouseout="reset_tooltip('summarizedexperiment-tooltip')"><span class="tooltiptext" id="summarizedexperiment-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>The figure below represents the anatomy of the SummarizedExperiment class.</p>
<p><img src="https://uclouvain-cbio.github.io/WSBIM1322/figs/SE.svg" width="80%"></p>
<p>Objects of the class SummarizedExperiment contain :</p>
<ul>
<li><p><strong>One (or more) assay(s)</strong> containing the quantitative omics data
(expression data), stored as a matrix-like object. Features (genes,
transcripts, proteins, …) are defined along the rows, and samples
along the columns.</p></li>
<li><p>A <strong>sample metadata</strong> slot containing sample co-variates, stored as a
data frame. Rows from this table represent samples (rows match exactly the
columns of the expression data).</p></li>
<li><p>A <strong>feature metadata</strong> slot containing feature co-variates, stored as
a data frame. The rows of this data frame match exactly the rows of the
expression data.</p></li>
</ul>
<p>The coordinated nature of the SummarizedExperiment guarantees that
during data manipulation, the dimensions of the different slots will
always match (i.e the columns in the expression data and then rows in
the sample metadata, as well as the rows in the expression data and
feature metadata) during data manipulation. For example, if we had to
exclude one sample from the assay, it would be automatically removed
from the sample metadata in the same operation.</p>
<p>The metadata slots can grow additional co-variates
(columns) without affecting the other structures.</p>
<div id="creating-a-summarizedexperiment" class="section level3" number="2.1.1">
<h3>
<span class="header-section-number">2.1.1</span> Creating a SummarizedExperiment<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('creating-a-summarizedexperiment')" onmouseout="reset_tooltip('creating-a-summarizedexperiment-tooltip')"><span class="tooltiptext" id="creating-a-summarizedexperiment-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>In order to create a <code>SummarizedExperiment</code>, we will create the
individual components, i.e the count matrix, the sample and gene
metadata from csv files. These are typically how RNA-Seq data are
provided (after raw data have been processed).</p>
<ul>
<li>
<strong>An expression matrix</strong>: we load the count matrix, specifying that
the first columns contains row/gene names, and convert the
<code>data.frame</code> to a <code>matrix</code>. You can download it
<a href="https://raw.githubusercontent.com/carpentries-incubator/bioc-intro/main/_episodes_rmd/data/count_matrix.csv">here</a>.</li>
</ul>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="sec-se.html#cb91-1" tabindex="-1"></a>count_matrix <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/count_matrix.csv"</span>,</span>
<span id="cb91-2"><a href="sec-se.html#cb91-2" tabindex="-1"></a>                         <span class="at">row.names =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb91-3"><a href="sec-se.html#cb91-3" tabindex="-1"></a>    <span class="fu">as.matrix</span>()</span>
<span id="cb91-4"><a href="sec-se.html#cb91-4" tabindex="-1"></a></span>
<span id="cb91-5"><a href="sec-se.html#cb91-5" tabindex="-1"></a>count_matrix[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<pre><code>##         GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
## Asl           1170        361        400        586        626        988
## Apod         36194      10347       9173      10620      13021      29594
## Cyp2d22       4060       1616       1603       1901       2171       3349
## Klk6           287        629        641        578        448        195
##         GSM2545342 GSM2545343 GSM2545344 GSM2545345 GSM2545346 GSM2545347
## Asl            836        535        586        597        938       1035
## Apod         24959      13668      13230      15868      27769      34301
## Cyp2d22       3122       2008       2254       2277       2985       3452
## Klk6           186       1101        537        567        327        233
##         GSM2545348 GSM2545349 GSM2545350 GSM2545351 GSM2545352 GSM2545353
## Asl            494        481        666        937        803        541
## Apod         11258      11812      15816      29242      20415      13682
## Cyp2d22       1883       2014       2417       3678       2920       2216
## Klk6           742        881        828        250        798        710
##         GSM2545354 GSM2545362 GSM2545363 GSM2545380
## Asl            473        748        576       1192
## Apod         11088      15916      11166      38148
## Cyp2d22       1821       2842       2011       4019
## Klk6           894        501        598        259
##  [ reached getOption("max.print") -- omitted 1 row ]</code></pre>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="sec-se.html#cb93-1" tabindex="-1"></a><span class="fu">dim</span>(count_matrix)</span></code></pre></div>
<pre><code>## [1] 1474   22</code></pre>
<ul>
<li>
<strong>A table describing the samples</strong>, available
<a href="https://raw.githubusercontent.com/carpentries-incubator/bioc-intro/main/_episodes_rmd/data/sample_metadata.csv">here</a>.</li>
</ul>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="sec-se.html#cb95-1" tabindex="-1"></a>sample_metadata <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/sample_metadata.csv"</span>)</span>
<span id="cb95-2"><a href="sec-se.html#cb95-2" tabindex="-1"></a>sample_metadata</span></code></pre></div>
<pre><code>##        sample     organism age    sex   infection  strain time     tissue mouse
## 1  GSM2545336 Mus musculus   8 Female  InfluenzaA C57BL/6    8 Cerebellum    14
## 2  GSM2545337 Mus musculus   8 Female NonInfected C57BL/6    0 Cerebellum     9
## 3  GSM2545338 Mus musculus   8 Female NonInfected C57BL/6    0 Cerebellum    10
## 4  GSM2545339 Mus musculus   8 Female  InfluenzaA C57BL/6    4 Cerebellum    15
## 5  GSM2545340 Mus musculus   8   Male  InfluenzaA C57BL/6    4 Cerebellum    18
## 6  GSM2545341 Mus musculus   8   Male  InfluenzaA C57BL/6    8 Cerebellum     6
## 7  GSM2545342 Mus musculus   8 Female  InfluenzaA C57BL/6    8 Cerebellum     5
## 8  GSM2545343 Mus musculus   8   Male NonInfected C57BL/6    0 Cerebellum    11
## 9  GSM2545344 Mus musculus   8 Female  InfluenzaA C57BL/6    4 Cerebellum    22
## 10 GSM2545345 Mus musculus   8   Male  InfluenzaA C57BL/6    4 Cerebellum    13
## 11 GSM2545346 Mus musculus   8   Male  InfluenzaA C57BL/6    8 Cerebellum    23
##  [ reached 'max' / getOption("max.print") -- omitted 11 rows ]</code></pre>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="sec-se.html#cb97-1" tabindex="-1"></a><span class="fu">dim</span>(sample_metadata)</span></code></pre></div>
<pre><code>## [1] 22  9</code></pre>
<ul>
<li>
<strong>A table describing the genes</strong>, available
<a href="https://raw.githubusercontent.com/carpentries-incubator/bioc-intro/main/_episodes_rmd/data/gene_metadata.csv">here</a>.</li>
</ul>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="sec-se.html#cb99-1" tabindex="-1"></a>gene_metadata <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/gene_metadata.csv"</span>)</span>
<span id="cb99-2"><a href="sec-se.html#cb99-2" tabindex="-1"></a>gene_metadata[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span></code></pre></div>
<pre><code>##       gene ENTREZID
## 1      Asl   109900
## 2     Apod    11815
## 3  Cyp2d22    56448
## 4     Klk6    19144
## 5    Fcrls    80891
## 6   Slc2a4    20528
## 7     Exd2    97827
## 8     Gjc2   118454
## 9     Plp1    18823
## 10    Gnb4    14696
##                                                                        product
## 1                                argininosuccinate lyase transcript variant X1
## 2                                        apolipoprotein D transcript variant 3
## 3     cytochrome P450 family 2 subfamily d polypeptide 22 transcript variant 2
## 4                          kallikrein related-peptidase 6 transcript variant 2
## 5                  Fc receptor-like S scavenger receptor transcript variant X1
## 6           solute carrier family 2 (facilitated glucose transporter) member 4
## 7                                        exonuclease 3'-5' domain containing 2
## 8                            gap junction protein gamma 2 transcript variant 1
## 9                          proteolipid protein (myelin) 1 transcript variant 1
## 10 guanine nucleotide binding protein (G protein) beta 4 transcript variant X2
##       ensembl_gene_id
## 1  ENSMUSG00000025533
## 2  ENSMUSG00000022548
## 3  ENSMUSG00000061740
## 4  ENSMUSG00000050063
## 5  ENSMUSG00000015852
## 6  ENSMUSG00000018566
## 7  ENSMUSG00000032705
## 8  ENSMUSG00000043448
## 9  ENSMUSG00000031425
## 10 ENSMUSG00000027669</code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="sec-se.html#cb101-1" tabindex="-1"></a><span class="fu">dim</span>(gene_metadata)</span></code></pre></div>
<pre><code>## [1] 1474    9</code></pre>
<p>We will create a <code>SummarizedExperiment</code> from these tables:</p>
<ul>
<li><p>The count matrix that will be used as the <strong><code>assay</code></strong></p></li>
<li><p>The table describing the samples will be used as the <strong>sample
metadata</strong> slot</p></li>
<li><p>The table describing the genes will be used as the <strong>features
metadata</strong> slot</p></li>
</ul>
<p>To do this we can put the different parts together using the
<code>SummarizedExperiment</code> constructor:</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="sec-se.html#cb103-1" tabindex="-1"></a><span class="do">## BiocManager::install("SummarizedExperiment")</span></span>
<span id="cb103-2"><a href="sec-se.html#cb103-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"SummarizedExperiment"</span>)</span></code></pre></div>
<p>First, we make sure that the samples are in the same order in the
count matrix and the sample annotation, and the same for the genes in
the count matrix and the gene annotation.</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="sec-se.html#cb104-1" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">rownames</span>(count_matrix) <span class="sc">==</span> gene_metadata<span class="sc">$</span>gene)</span>
<span id="cb104-2"><a href="sec-se.html#cb104-2" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">colnames</span>(count_matrix) <span class="sc">==</span> sample_metadata<span class="sc">$</span>sample)</span></code></pre></div>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="sec-se.html#cb105-1" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(<span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> count_matrix),</span>
<span id="cb105-2"><a href="sec-se.html#cb105-2" tabindex="-1"></a>                           <span class="at">colData =</span> sample_metadata,</span>
<span id="cb105-3"><a href="sec-se.html#cb105-3" tabindex="-1"></a>                           <span class="at">rowData =</span> gene_metadata)</span>
<span id="cb105-4"><a href="sec-se.html#cb105-4" tabindex="-1"></a>se</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 1474 22 
## metadata(0):
## assays(1): counts
## rownames(1474): Asl Apod ... Pbx1 Rgs4
## rowData names(9): gene ENTREZID ... phenotype_description
##   hsapiens_homolog_associated_gene_name
## colnames(22): GSM2545336 GSM2545337 ... GSM2545363 GSM2545380
## colData names(9): sample organism ... tissue mouse</code></pre>
<p>Using this data structure, we can access the expression matrix with
the <code>assay</code> function:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="sec-se.html#cb107-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">assay</span>(se))</span></code></pre></div>
<pre><code>##         GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
## Asl           1170        361        400        586        626        988
## Apod         36194      10347       9173      10620      13021      29594
## Cyp2d22       4060       1616       1603       1901       2171       3349
## Klk6           287        629        641        578        448        195
##         GSM2545342 GSM2545343 GSM2545344 GSM2545345 GSM2545346 GSM2545347
## Asl            836        535        586        597        938       1035
## Apod         24959      13668      13230      15868      27769      34301
## Cyp2d22       3122       2008       2254       2277       2985       3452
## Klk6           186       1101        537        567        327        233
##         GSM2545348 GSM2545349 GSM2545350 GSM2545351 GSM2545352 GSM2545353
## Asl            494        481        666        937        803        541
## Apod         11258      11812      15816      29242      20415      13682
## Cyp2d22       1883       2014       2417       3678       2920       2216
## Klk6           742        881        828        250        798        710
##         GSM2545354 GSM2545362 GSM2545363 GSM2545380
## Asl            473        748        576       1192
## Apod         11088      15916      11166      38148
## Cyp2d22       1821       2842       2011       4019
## Klk6           894        501        598        259
##  [ reached getOption("max.print") -- omitted 2 rows ]</code></pre>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="sec-se.html#cb109-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">assay</span>(se))</span></code></pre></div>
<pre><code>## [1] 1474   22</code></pre>
<p>We can access the sample metadata using the <code>colData</code> function:</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="sec-se.html#cb111-1" tabindex="-1"></a><span class="fu">colData</span>(se)</span></code></pre></div>
<pre><code>## DataFrame with 22 rows and 9 columns
##                 sample     organism       age         sex   infection
##            &lt;character&gt;  &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## GSM2545336  GSM2545336 Mus musculus         8      Female  InfluenzaA
## GSM2545337  GSM2545337 Mus musculus         8      Female NonInfected
## GSM2545338  GSM2545338 Mus musculus         8      Female NonInfected
## GSM2545339  GSM2545339 Mus musculus         8      Female  InfluenzaA
## GSM2545340  GSM2545340 Mus musculus         8        Male  InfluenzaA
## ...                ...          ...       ...         ...         ...
## GSM2545353  GSM2545353 Mus musculus         8      Female NonInfected
## GSM2545354  GSM2545354 Mus musculus         8        Male NonInfected
## GSM2545362  GSM2545362 Mus musculus         8      Female  InfluenzaA
## GSM2545363  GSM2545363 Mus musculus         8        Male  InfluenzaA
##                 strain      time      tissue     mouse
##            &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
## GSM2545336     C57BL/6         8  Cerebellum        14
## GSM2545337     C57BL/6         0  Cerebellum         9
## GSM2545338     C57BL/6         0  Cerebellum        10
## GSM2545339     C57BL/6         4  Cerebellum        15
## GSM2545340     C57BL/6         4  Cerebellum        18
## ...                ...       ...         ...       ...
## GSM2545353     C57BL/6         0  Cerebellum         4
## GSM2545354     C57BL/6         0  Cerebellum         2
## GSM2545362     C57BL/6         4  Cerebellum        20
## GSM2545363     C57BL/6         4  Cerebellum        12
##  [ reached getOption("max.print") -- omitted 1 row ]</code></pre>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="sec-se.html#cb113-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">colData</span>(se))</span></code></pre></div>
<pre><code>## [1] 22  9</code></pre>
<p>We can also access the feature metadata using the <code>rowData</code> function:</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="sec-se.html#cb115-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(se))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 9 columns
##                gene  ENTREZID                product    ensembl_gene_id
##         &lt;character&gt; &lt;integer&gt;            &lt;character&gt;        &lt;character&gt;
## Asl             Asl    109900 argininosuccinate ly.. ENSMUSG00000025533
## Apod           Apod     11815 apolipoprotein D tra.. ENSMUSG00000022548
## Cyp2d22     Cyp2d22     56448 cytochrome P450 fami.. ENSMUSG00000061740
## Klk6           Klk6     19144 kallikrein related-p.. ENSMUSG00000050063
## Fcrls         Fcrls     80891 Fc receptor-like S s.. ENSMUSG00000015852
## Slc2a4       Slc2a4     20528 solute carrier famil.. ENSMUSG00000018566
##         external_synonym chromosome_name   gene_biotype  phenotype_description
##              &lt;character&gt;     &lt;character&gt;    &lt;character&gt;            &lt;character&gt;
## Asl        2510006M18Rik               5 protein_coding abnormal circulating..
## Apod                  NA              16 protein_coding abnormal lipid homeo..
## Cyp2d22             2D22              15 protein_coding abnormal skin morpho..
## Klk6                Bssp               7 protein_coding abnormal cytokine le..
## Fcrls      2810439C17Rik               3 protein_coding decreased CD8-positi..
## Slc2a4            Glut-4              11 protein_coding abnormal circulating..
##         hsapiens_homolog_associated_gene_name
##                                   &lt;character&gt;
## Asl                                       ASL
## Apod                                     APOD
## Cyp2d22                                CYP2D6
## Klk6                                     KLK6
## Fcrls                                   FCRL4
## Slc2a4                                 SLC2A4</code></pre>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="sec-se.html#cb117-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">rowData</span>(se))</span></code></pre></div>
<pre><code>## [1] 1474    9</code></pre>
</div>
<div id="subsetting-a-summarizedexperiment" class="section level3" number="2.1.2">
<h3>
<span class="header-section-number">2.1.2</span> Subsetting a SummarizedExperiment<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('subsetting-a-summarizedexperiment')" onmouseout="reset_tooltip('subsetting-a-summarizedexperiment-tooltip')"><span class="tooltiptext" id="subsetting-a-summarizedexperiment-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h3>
<p>SummarizedExperiment can be subset just like with data frames, with
numerics or with characters of logicals.</p>
<p>Below, we create a new instance of class SummarizedExperiment that
contains only the 5 first features for the 3 first samples.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="sec-se.html#cb119-1" tabindex="-1"></a>se1 <span class="ot">&lt;-</span> se[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb119-2"><a href="sec-se.html#cb119-2" tabindex="-1"></a>se1</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 5 3 
## metadata(0):
## assays(1): counts
## rownames(5): Asl Apod Cyp2d22 Klk6 Fcrls
## rowData names(9): gene ENTREZID ... phenotype_description
##   hsapiens_homolog_associated_gene_name
## colnames(3): GSM2545336 GSM2545337 GSM2545338
## colData names(9): sample organism ... tissue mouse</code></pre>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="sec-se.html#cb121-1" tabindex="-1"></a><span class="fu">colData</span>(se1)</span></code></pre></div>
<pre><code>## DataFrame with 3 rows and 9 columns
##                 sample     organism       age         sex   infection
##            &lt;character&gt;  &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## GSM2545336  GSM2545336 Mus musculus         8      Female  InfluenzaA
## GSM2545337  GSM2545337 Mus musculus         8      Female NonInfected
## GSM2545338  GSM2545338 Mus musculus         8      Female NonInfected
##                 strain      time      tissue     mouse
##            &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
## GSM2545336     C57BL/6         8  Cerebellum        14
## GSM2545337     C57BL/6         0  Cerebellum         9
## GSM2545338     C57BL/6         0  Cerebellum        10</code></pre>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="sec-se.html#cb123-1" tabindex="-1"></a><span class="fu">rowData</span>(se1)</span></code></pre></div>
<pre><code>## DataFrame with 5 rows and 9 columns
##                gene  ENTREZID                product    ensembl_gene_id
##         &lt;character&gt; &lt;integer&gt;            &lt;character&gt;        &lt;character&gt;
## Asl             Asl    109900 argininosuccinate ly.. ENSMUSG00000025533
## Apod           Apod     11815 apolipoprotein D tra.. ENSMUSG00000022548
## Cyp2d22     Cyp2d22     56448 cytochrome P450 fami.. ENSMUSG00000061740
## Klk6           Klk6     19144 kallikrein related-p.. ENSMUSG00000050063
## Fcrls         Fcrls     80891 Fc receptor-like S s.. ENSMUSG00000015852
##         external_synonym chromosome_name   gene_biotype  phenotype_description
##              &lt;character&gt;     &lt;character&gt;    &lt;character&gt;            &lt;character&gt;
## Asl        2510006M18Rik               5 protein_coding abnormal circulating..
## Apod                  NA              16 protein_coding abnormal lipid homeo..
## Cyp2d22             2D22              15 protein_coding abnormal skin morpho..
## Klk6                Bssp               7 protein_coding abnormal cytokine le..
## Fcrls      2810439C17Rik               3 protein_coding decreased CD8-positi..
##         hsapiens_homolog_associated_gene_name
##                                   &lt;character&gt;
## Asl                                       ASL
## Apod                                     APOD
## Cyp2d22                                CYP2D6
## Klk6                                     KLK6
## Fcrls                                   FCRL4</code></pre>
<p>We can also use the <code>colData()</code> function to subset on something from
the sample metadata or the <code>rowData()</code> to subset on something from the
feature metadata.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Create a new object keeping only miRNAs (see <code>gene_biotype</code> feature
variable) and the non infected samples.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-7" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-7', 'sol-start-7')"></span>
</p>
<div id="sol-body-7" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="sec-se.html#cb125-1" tabindex="-1"></a>se1 <span class="ot">&lt;-</span> se[<span class="fu">rowData</span>(se)<span class="sc">$</span>gene_biotype <span class="sc">==</span> <span class="st">"miRNA"</span>,</span>
<span id="cb125-2"><a href="sec-se.html#cb125-2" tabindex="-1"></a>          <span class="fu">colData</span>(se)<span class="sc">$</span>infection <span class="sc">==</span> <span class="st">"NonInfected"</span>]</span>
<span id="cb125-3"><a href="sec-se.html#cb125-3" tabindex="-1"></a>se1</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 7 7 
## metadata(0):
## assays(1): counts
## rownames(7): Mir1901 Mir378a ... Mir128-1 Mir7682
## rowData names(9): gene ENTREZID ... phenotype_description
##   hsapiens_homolog_associated_gene_name
## colnames(7): GSM2545337 GSM2545338 ... GSM2545353 GSM2545354
## colData names(9): sample organism ... tissue mouse</code></pre>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="sec-se.html#cb127-1" tabindex="-1"></a><span class="fu">assay</span>(se1)</span></code></pre></div>
<pre><code>##          GSM2545337 GSM2545338 GSM2545343 GSM2545348 GSM2545349 GSM2545353
## Mir1901          45         44         74         55         68         33
## Mir378a          11          7          9          4         12          4
## Mir133b           4          6          5          4          6          7
## Mir30c-2         10          6         16         12          8         17
## Mir149            1          2          0          0          0          0
## Mir128-1          4          1          2          2          1          2
## Mir7682           2          0          4          1          3          5
##          GSM2545354
## Mir1901          60
## Mir378a           8
## Mir133b           3
## Mir30c-2         15
## Mir149            2
## Mir128-1          1
## Mir7682           5</code></pre>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="sec-se.html#cb129-1" tabindex="-1"></a><span class="fu">colData</span>(se1)</span></code></pre></div>
<pre><code>## DataFrame with 7 rows and 9 columns
##                 sample     organism       age         sex   infection
##            &lt;character&gt;  &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## GSM2545337  GSM2545337 Mus musculus         8      Female NonInfected
## GSM2545338  GSM2545338 Mus musculus         8      Female NonInfected
## GSM2545343  GSM2545343 Mus musculus         8        Male NonInfected
## GSM2545348  GSM2545348 Mus musculus         8      Female NonInfected
## GSM2545349  GSM2545349 Mus musculus         8        Male NonInfected
## GSM2545353  GSM2545353 Mus musculus         8      Female NonInfected
## GSM2545354  GSM2545354 Mus musculus         8        Male NonInfected
##                 strain      time      tissue     mouse
##            &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
## GSM2545337     C57BL/6         0  Cerebellum         9
## GSM2545338     C57BL/6         0  Cerebellum        10
## GSM2545343     C57BL/6         0  Cerebellum        11
## GSM2545348     C57BL/6         0  Cerebellum         8
## GSM2545349     C57BL/6         0  Cerebellum         7
## GSM2545353     C57BL/6         0  Cerebellum         4
## GSM2545354     C57BL/6         0  Cerebellum         2</code></pre>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="sec-se.html#cb131-1" tabindex="-1"></a><span class="fu">rowData</span>(se1)</span></code></pre></div>
<pre><code>## DataFrame with 7 rows and 9 columns
##                 gene  ENTREZID        product    ensembl_gene_id
##          &lt;character&gt; &lt;integer&gt;    &lt;character&gt;        &lt;character&gt;
## Mir1901      Mir1901 100316686  microRNA 1901 ENSMUSG00000084565
## Mir378a      Mir378a    723889  microRNA 378a ENSMUSG00000105200
## Mir133b      Mir133b    723817  microRNA 133b ENSMUSG00000065480
## Mir30c-2    Mir30c-2    723964 microRNA 30c-2 ENSMUSG00000065567
## Mir149        Mir149    387167   microRNA 149 ENSMUSG00000065470
## Mir128-1    Mir128-1    387147 microRNA 128-1 ENSMUSG00000065520
## Mir7682      Mir7682 102466847  microRNA 7682 ENSMUSG00000106406
##          external_synonym chromosome_name gene_biotype  phenotype_description
##               &lt;character&gt;     &lt;character&gt;  &lt;character&gt;            &lt;character&gt;
## Mir1901          Mirn1901              18        miRNA                     NA
## Mir378a           Mirn378              18        miRNA abnormal mitochondri..
## Mir133b          mir 133b               1        miRNA no abnormal phenotyp..
## Mir30c-2        mir 30c-2               1        miRNA                     NA
## Mir149            Mirn149               1        miRNA increased circulatin..
## Mir128-1          Mirn128               1        miRNA no abnormal phenotyp..
## Mir7682      mmu-mir-7682               1        miRNA                     NA
##          hsapiens_homolog_associated_gene_name
##                                    &lt;character&gt;
## Mir1901                                     NA
## Mir378a                                MIR378A
## Mir133b                                MIR133B
## Mir30c-2                               MIR30C2
## Mir149                                      NA
## Mir128-1                              MIR128-1
## Mir7682                                     NA</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Extract the gene expression levels of the 3 first genes in samples at
time 0 and at time 8.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-8" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-8', 'sol-start-8')"></span>
</p>
<div id="sol-body-8" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="sec-se.html#cb133-1" tabindex="-1"></a><span class="fu">assay</span>(se)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="fu">colData</span>(se)<span class="sc">$</span>time <span class="sc">!=</span> <span class="dv">4</span>]</span></code></pre></div>
<pre><code>##         GSM2545336 GSM2545337 GSM2545338 GSM2545341 GSM2545342 GSM2545343
## Asl           1170        361        400        988        836        535
## Apod         36194      10347       9173      29594      24959      13668
## Cyp2d22       4060       1616       1603       3349       3122       2008
##         GSM2545346 GSM2545347 GSM2545348 GSM2545349 GSM2545351 GSM2545353
## Asl            938       1035        494        481        937        541
## Apod         27769      34301      11258      11812      29242      13682
## Cyp2d22       2985       3452       1883       2014       3678       2216
##         GSM2545354 GSM2545380
## Asl            473       1192
## Apod         11088      38148
## Cyp2d22       1821       4019</code></pre>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="sec-se.html#cb135-1" tabindex="-1"></a><span class="co"># Equivalent to</span></span>
<span id="cb135-2"><a href="sec-se.html#cb135-2" tabindex="-1"></a><span class="fu">assay</span>(se)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="fu">colData</span>(se)<span class="sc">$</span>time <span class="sc">==</span> <span class="dv">0</span> <span class="sc">|</span> <span class="fu">colData</span>(se)<span class="sc">$</span>time <span class="sc">==</span> <span class="dv">8</span>]</span></code></pre></div>
<pre><code>##         GSM2545336 GSM2545337 GSM2545338 GSM2545341 GSM2545342 GSM2545343
## Asl           1170        361        400        988        836        535
## Apod         36194      10347       9173      29594      24959      13668
## Cyp2d22       4060       1616       1603       3349       3122       2008
##         GSM2545346 GSM2545347 GSM2545348 GSM2545349 GSM2545351 GSM2545353
## Asl            938       1035        494        481        937        541
## Apod         27769      34301      11258      11812      29242      13682
## Cyp2d22       2985       3452       1883       2014       3678       2216
##         GSM2545354 GSM2545380
## Asl            473       1192
## Apod         11088      38148
## Cyp2d22       1821       4019</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div id="adding-variables-to-metadata" class="section level4" number="2.1.2.1">
<h4>
<span class="header-section-number">2.1.2.1</span> Adding variables to metadata</h4>
<p>We can also add information to the metadata. Suppose that you want to
add the center where the samples were collected …</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="sec-se.html#cb137-1" tabindex="-1"></a><span class="fu">colData</span>(se)<span class="sc">$</span>center <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"University of Illinois"</span>, <span class="fu">nrow</span>(<span class="fu">colData</span>(se)))</span>
<span id="cb137-2"><a href="sec-se.html#cb137-2" tabindex="-1"></a><span class="fu">colData</span>(se)</span></code></pre></div>
<pre><code>## DataFrame with 22 rows and 10 columns
##                 sample     organism       age         sex   infection
##            &lt;character&gt;  &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## GSM2545336  GSM2545336 Mus musculus         8      Female  InfluenzaA
## GSM2545337  GSM2545337 Mus musculus         8      Female NonInfected
## GSM2545338  GSM2545338 Mus musculus         8      Female NonInfected
## GSM2545339  GSM2545339 Mus musculus         8      Female  InfluenzaA
## GSM2545340  GSM2545340 Mus musculus         8        Male  InfluenzaA
## ...                ...          ...       ...         ...         ...
## GSM2545353  GSM2545353 Mus musculus         8      Female NonInfected
## GSM2545354  GSM2545354 Mus musculus         8        Male NonInfected
## GSM2545362  GSM2545362 Mus musculus         8      Female  InfluenzaA
##                 strain      time      tissue     mouse                 center
##            &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;integer&gt;            &lt;character&gt;
## GSM2545336     C57BL/6         8  Cerebellum        14 University of Illinois
## GSM2545337     C57BL/6         0  Cerebellum         9 University of Illinois
## GSM2545338     C57BL/6         0  Cerebellum        10 University of Illinois
## GSM2545339     C57BL/6         4  Cerebellum        15 University of Illinois
## GSM2545340     C57BL/6         4  Cerebellum        18 University of Illinois
## ...                ...       ...         ...       ...                    ...
## GSM2545353     C57BL/6         0  Cerebellum         4 University of Illinois
## GSM2545354     C57BL/6         0  Cerebellum         2 University of Illinois
## GSM2545362     C57BL/6         4  Cerebellum        20 University of Illinois
##  [ reached getOption("max.print") -- omitted 2 rows ]</code></pre>
<p>This illustrates that the metadata slots can grow indefinitely without
affecting the other structures!</p>
</div>
</div>
</div>
<div id="rangedsummarizedexperiment" class="section level2" number="2.2">
<h2>
<span class="header-section-number">2.2</span> RangedSummarizedExperiment<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('rangedsummarizedexperiment')" onmouseout="reset_tooltip('rangedsummarizedexperiment-tooltip')"><span class="tooltiptext" id="rangedsummarizedexperiment-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>In addition to the <code>SummarizedExperiment</code> slots we have seen above, it
is also possible to store information about the features’ genomic
ranges. Below, we load the <code>airway</code> <code>RangedSummarizedExperiment</code> data
to illustrate this:</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="sec-se.html#cb139-1" tabindex="-1"></a><span class="fu">library</span>(airway)</span>
<span id="cb139-2"><a href="sec-se.html#cb139-2" tabindex="-1"></a><span class="fu">data</span>(airway)</span>
<span id="cb139-3"><a href="sec-se.html#cb139-3" tabindex="-1"></a>airway</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 63677 8 
## metadata(1): ''
## assays(1): counts
## rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492
##   ENSG00000273493
## rowData names(10): gene_id gene_name ... seq_coord_system symbol
## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
## colData names(9): SampleName cell ... Sample BioSample</code></pre>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="sec-se.html#cb141-1" tabindex="-1"></a><span class="fu">rowRanges</span>(airway)</span></code></pre></div>
<pre><code>## GRangesList object of length 63677:
## $ENSG00000000003
## GRanges object with 17 ranges and 2 metadata columns:
##        seqnames            ranges strand |   exon_id       exon_name
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt;
##    [1]        X 99883667-99884983      - |    667145 ENSE00001459322
##    [2]        X 99885756-99885863      - |    667146 ENSE00000868868
##    [3]        X 99887482-99887565      - |    667147 ENSE00000401072
##    [4]        X 99887538-99887565      - |    667148 ENSE00001849132
##    [5]        X 99888402-99888536      - |    667149 ENSE00003554016
##    ...      ...               ...    ... .       ...             ...
##   [13]        X 99890555-99890743      - |    667156 ENSE00003512331
##   [14]        X 99891188-99891686      - |    667158 ENSE00001886883
##   [15]        X 99891605-99891803      - |    667159 ENSE00001855382
##   [16]        X 99891790-99892101      - |    667160 ENSE00001863395
##   [17]        X 99894942-99894988      - |    667161 ENSE00001828996
##   -------
##   seqinfo: 722 sequences (1 circular) from an unspecified genome
## 
## ...
## &lt;63676 more elements&gt;</code></pre>
<p>This <code>GRangesList</code> contains a list of 64102 range elements, one for
each gene/feature. Each element is a <code>GRanges</code> object that contains
the positions of the exons or the genes:</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="sec-se.html#cb143-1" tabindex="-1"></a><span class="fu">rowRanges</span>(airway)[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## GRanges object with 17 ranges and 2 metadata columns:
##        seqnames            ranges strand |   exon_id       exon_name
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt;
##    [1]        X 99883667-99884983      - |    667145 ENSE00001459322
##    [2]        X 99885756-99885863      - |    667146 ENSE00000868868
##    [3]        X 99887482-99887565      - |    667147 ENSE00000401072
##    [4]        X 99887538-99887565      - |    667148 ENSE00001849132
##    [5]        X 99888402-99888536      - |    667149 ENSE00003554016
##    ...      ...               ...    ... .       ...             ...
##   [13]        X 99890555-99890743      - |    667156 ENSE00003512331
##   [14]        X 99891188-99891686      - |    667158 ENSE00001886883
##   [15]        X 99891605-99891803      - |    667159 ENSE00001855382
##   [16]        X 99891790-99892101      - |    667160 ENSE00001863395
##   [17]        X 99894942-99894988      - |    667161 ENSE00001828996
##   -------
##   seqinfo: 722 sequences (1 circular) from an unspecified genome</code></pre>
<p>It is possible to use these row genomic ranges to subset features in
regions of interest, defined as <code>GRanges</code> object. Below, for example,
we subset features between bases 1M and 2M on chromosome 7:</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="sec-se.html#cb145-1" tabindex="-1"></a>roi <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">"7"</span>, <span class="at">ranges =</span> <span class="fl">1e6</span><span class="sc">:</span><span class="fl">2e6</span>)</span>
<span id="cb145-2"><a href="sec-se.html#cb145-2" tabindex="-1"></a><span class="fu">subsetByOverlaps</span>(airway, roi)</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 34 8 
## metadata(1): ''
## assays(1): counts
## rownames(34): ENSG00000002822 ENSG00000073067 ... ENSG00000266391
##   ENSG00000273230
## rowData names(10): gene_id gene_name ... seq_coord_system symbol
## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
## colData names(9): SampleName cell ... Sample BioSample</code></pre>
</div>
<div id="creating-a-count-matrix-from-multiple-files" class="section level2" number="2.3">
<h2>
<span class="header-section-number">2.3</span> Creating a count matrix from multiple files<div class="tooltip"><button class="internal-link-btn" onclick="copy_link('creating-a-count-matrix-from-multiple-files')" onmouseout="reset_tooltip('creating-a-count-matrix-from-multiple-files-tooltip')"><span class="tooltiptext" id="creating-a-count-matrix-from-multiple-files-tooltip">Copy link</span><i class="fa fa-link"></i></button></div>
</h2>
<p>Raw HTS fastq files from different sample are processed into raw
counts one by one (see next chapter). These data meed to be merged in
a single count matrix, were each column represents a sample, and each
line represents a gene. This count matrix will be the starting point
of a differential expression analysis.</p>
<pre><code>##                 sample1 sample2 sample3 sample4 sample5 sample6
## ENSG00000223972       0       0       0       0       0       1
## ENSG00000227232      14      28      17      40      16      13
## ENSG00000278267       8       4       3       1       1       6
## ENSG00000243485       0       0       0       0       0       0
## ENSG00000284332       0       0       0       0       0       0
## ENSG00000237613       0       0       0       0       0       0</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Imagine you had initially 6 samples (3 control and 3 treated samples)
that you have processed independently giving the 6 counts files
generated by FeatureCounts located in the folder
<code>wsbim_data/count_data/</code>.</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="sec-se.html#cb148-1" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"wsbim2122_data/count_data"</span>,</span>
<span id="cb148-2"><a href="sec-se.html#cb148-2" tabindex="-1"></a>                      <span class="at">pattern =</span> <span class="st">"*tsv.gz"</span>,</span>
<span id="cb148-3"><a href="sec-se.html#cb148-3" tabindex="-1"></a>                      <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb148-4"><a href="sec-se.html#cb148-4" tabindex="-1"></a>samples</span></code></pre></div>
<pre><code>## [1] "wsbim2122_data/count_data/sample1_counts.tsv.gz"
## [2] "wsbim2122_data/count_data/sample2_counts.tsv.gz"
## [3] "wsbim2122_data/count_data/sample3_counts.tsv.gz"
## [4] "wsbim2122_data/count_data/sample4_counts.tsv.gz"
## [5] "wsbim2122_data/count_data/sample5_counts.tsv.gz"
## [6] "wsbim2122_data/count_data/sample6_counts.tsv.gz"</code></pre>
<p>Use these files to create a count matrix.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="msmb-solution">
<p class="solution-begin">
► Solution<span id="sol-start-9" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-9', 'sol-start-9')"></span>
</p>
<div id="sol-body-9" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="sec-se.html#cb150-1" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(samples[<span class="dv">1</span>]) <span class="sc">%&gt;%</span></span>
<span id="cb150-2"><a href="sec-se.html#cb150-2" tabindex="-1"></a>    <span class="fu">select</span>(Geneid, <span class="fu">ends_with</span>(<span class="st">'.bam'</span>))</span>
<span id="cb150-3"><a href="sec-se.html#cb150-3" tabindex="-1"></a></span>
<span id="cb150-4"><a href="sec-se.html#cb150-4" tabindex="-1"></a><span class="cf">for</span> (sample <span class="cf">in</span> samples[<span class="sc">-</span><span class="dv">1</span>]){</span>
<span id="cb150-5"><a href="sec-se.html#cb150-5" tabindex="-1"></a>  tmp <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(sample) <span class="sc">%&gt;%</span></span>
<span id="cb150-6"><a href="sec-se.html#cb150-6" tabindex="-1"></a>      <span class="fu">select</span>(<span class="fu">ends_with</span>(<span class="st">'.bam'</span>))</span>
<span id="cb150-7"><a href="sec-se.html#cb150-7" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> <span class="fu">cbind</span>(counts, tmp)</span>
<span id="cb150-8"><a href="sec-se.html#cb150-8" tabindex="-1"></a>}</span>
<span id="cb150-9"><a href="sec-se.html#cb150-9" tabindex="-1"></a></span>
<span id="cb150-10"><a href="sec-se.html#cb150-10" tabindex="-1"></a><span class="fu">names</span>(counts) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="at">pattern =</span> <span class="st">".bam$"</span>, <span class="st">''</span>, <span class="fu">names</span>(counts))</span>
<span id="cb150-11"><a href="sec-se.html#cb150-11" tabindex="-1"></a><span class="fu">names</span>(counts) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="at">pattern =</span> <span class="st">"../processed_data/bam/"</span>, <span class="st">''</span>, <span class="fu">names</span>(counts))</span>
<span id="cb150-12"><a href="sec-se.html#cb150-12" tabindex="-1"></a><span class="fu">head</span>(counts)</span></code></pre></div>
<pre><code>##            Geneid sample1 sample2 sample3 sample4 sample5 sample6
## 1 ENSG00000223972       0       0       0       0       0       1
## 2 ENSG00000227232      14      28      17      40      16      13
## 3 ENSG00000278267       8       4       3       1       1       6
## 4 ENSG00000243485       0       0       0       0       0       0
## 5 ENSG00000284332       0       0       0       0       0       0
## 6 ENSG00000237613       0       0       0       0       0       0</code></pre>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="sec-se.html#cb152-1" tabindex="-1"></a><span class="do">## We want the 7th column from each file</span></span>
<span id="cb152-2"><a href="sec-se.html#cb152-2" tabindex="-1"></a><span class="fu">sapply</span>(samples, <span class="cf">function</span>(x) <span class="fu">names</span>(<span class="fu">read.delim</span>(x)))</span></code></pre></div>
<pre><code>##      wsbim2122_data/count_data/sample1_counts.tsv.gz
## [1,] "Geneid"                                       
## [2,] "Chr"                                          
## [3,] "Start"                                        
## [4,] "End"                                          
## [5,] "Strand"                                       
## [6,] "Length"                                       
## [7,] "...processed_data.bam.sample1.bam"            
##      wsbim2122_data/count_data/sample2_counts.tsv.gz
## [1,] "Geneid"                                       
## [2,] "Chr"                                          
## [3,] "Start"                                        
## [4,] "End"                                          
## [5,] "Strand"                                       
## [6,] "Length"                                       
## [7,] "...processed_data.bam.sample2.bam"            
##      wsbim2122_data/count_data/sample3_counts.tsv.gz
## [1,] "Geneid"                                       
## [2,] "Chr"                                          
## [3,] "Start"                                        
## [4,] "End"                                          
## [5,] "Strand"                                       
## [6,] "Length"                                       
## [7,] "...processed_data.bam.sample3.bam"            
##      wsbim2122_data/count_data/sample4_counts.tsv.gz
## [1,] "Geneid"                                       
## [2,] "Chr"                                          
## [3,] "Start"                                        
## [4,] "End"                                          
## [5,] "Strand"                                       
## [6,] "Length"                                       
## [7,] "...processed_data.bam.sample4.bam"            
##      wsbim2122_data/count_data/sample5_counts.tsv.gz
## [1,] "Geneid"                                       
## [2,] "Chr"                                          
## [3,] "Start"                                        
## [4,] "End"                                          
## [5,] "Strand"                                       
## [6,] "Length"                                       
## [7,] "...processed_data.bam.sample5.bam"            
##      wsbim2122_data/count_data/sample6_counts.tsv.gz
## [1,] "Geneid"                                       
## [2,] "Chr"                                          
## [3,] "Start"                                        
## [4,] "End"                                          
## [5,] "Strand"                                       
## [6,] "Length"                                       
## [7,] "...processed_data.bam.sample6.bam"</code></pre>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="sec-se.html#cb154-1" tabindex="-1"></a><span class="do">## Create the counts matrix</span></span>
<span id="cb154-2"><a href="sec-se.html#cb154-2" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">sapply</span>(samples, <span class="cf">function</span>(x) <span class="fu">read.delim</span>(x)[, <span class="dv">7</span>])</span>
<span id="cb154-3"><a href="sec-se.html#cb154-3" tabindex="-1"></a></span>
<span id="cb154-4"><a href="sec-se.html#cb154-4" tabindex="-1"></a><span class="do">## Update row/col names</span></span>
<span id="cb154-5"><a href="sec-se.html#cb154-5" tabindex="-1"></a><span class="fu">rownames</span>(counts) <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(samples[<span class="dv">1</span>])[[<span class="dv">1</span>]]</span>
<span id="cb154-6"><a href="sec-se.html#cb154-6" tabindex="-1"></a><span class="fu">colnames</span>(counts) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(sample, <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(counts))</span>
<span id="cb154-7"><a href="sec-se.html#cb154-7" tabindex="-1"></a><span class="fu">head</span>(counts)</span></code></pre></div>
<pre><code>##                 wsbim2122_data/count_data/sample6_counts.tsv.gz1
## ENSG00000223972                                                0
## ENSG00000227232                                               14
## ENSG00000278267                                                8
## ENSG00000243485                                                0
## ENSG00000284332                                                0
## ENSG00000237613                                                0
##                 wsbim2122_data/count_data/sample6_counts.tsv.gz2
## ENSG00000223972                                                0
## ENSG00000227232                                               28
## ENSG00000278267                                                4
## ENSG00000243485                                                0
## ENSG00000284332                                                0
## ENSG00000237613                                                0
##                 wsbim2122_data/count_data/sample6_counts.tsv.gz3
## ENSG00000223972                                                0
## ENSG00000227232                                               17
## ENSG00000278267                                                3
## ENSG00000243485                                                0
## ENSG00000284332                                                0
## ENSG00000237613                                                0
##                 wsbim2122_data/count_data/sample6_counts.tsv.gz4
## ENSG00000223972                                                0
## ENSG00000227232                                               40
## ENSG00000278267                                                1
## ENSG00000243485                                                0
## ENSG00000284332                                                0
## ENSG00000237613                                                0
##                 wsbim2122_data/count_data/sample6_counts.tsv.gz5
## ENSG00000223972                                                0
## ENSG00000227232                                               16
## ENSG00000278267                                                1
## ENSG00000243485                                                0
## ENSG00000284332                                                0
## ENSG00000237613                                                0
##                 wsbim2122_data/count_data/sample6_counts.tsv.gz6
## ENSG00000223972                                                1
## ENSG00000227232                                               13
## ENSG00000278267                                                6
## ENSG00000243485                                                0
## ENSG00000284332                                                0
## ENSG00000237613                                                0</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>

</div>
</div>
<div class="footnotes">
<hr>
<ol start="5">
<li id="fn5"><p>The <a href="http://www.bioconductor.org">Bioconductor</a> was
initiated by Robert Gentleman, one of the two creators of the R
language. Bioconductor provides tools for the analysis and
comprehension of omics data. Bioconductor uses the R statistical
programming language, and is open source and open development.<a href="sec-se.html#fnref5" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
</body></html>

<p style="text-align: center;">
<a href="finding-differentially-expressed-features.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-dimred.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2023-10-03
 using 
R version 4.3.1 Patched (2023-07-10 r84676)
</p>
</div>
</div>



</body>
</html>
